Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm

```diff
@@ -458,22 +458,31 @@ static inline String toMIMEType(WebExtensionTab::ImageFormat format)
 
 void WebExtensionContext::tabsSendMessage(WebExtensionTabIdentifier tabIdentifier, const String& messageJSON, std::optional<WebExtensionFrameIdentifier> frameIdentifier, const WebExtensionMessageSenderParameters& senderParameters, CompletionHandler<void(Expected<String, WebExtensionError>&&)>&& completionHandler)
 {
+    static NSString * const apiName = @"tabs.sendMessage()";
+
     RefPtr tab = getTab(tabIdentifier);
     if (!tab) {
-        completionHandler(toWebExtensionError(@"tabs.sendMessage()", nil, @"tab not found"));
+        completionHandler(toWebExtensionError(apiName, nil, @"tab not found"));
+        return;
+    }
+
+    if (!tab->extensionHasPermission()) {
+        completionHandler(toWebExtensionError(apiName, nil, @"this extension does not have access to this tab"));
         return;
     }
 
-    auto contentScriptProcesses = tab->processes(WebExtensionEventListenerType::RuntimeOnMessage, WebExtensionContentWorldType::ContentScript);
-    if (contentScriptProcesses.isEmpty()) {
+    auto worldType = isURLForThisExtension(tab->url()) ? WebExtensionContentWorldType::Main : WebExtensionContentWorldType::ContentScript;
```

Comment:
`targetContentWorldType` would be consistent with other code, since there are a lot of worlds involved in messaging.